frontend/lib/resolution/resolution-types.cpp

```diff
@@ -46,6 +46,32 @@ namespace resolution {
 using namespace uast;
 using namespace types;
 
+bool
+UntypedFnSignature::FormalDetail::operator<(const FormalDetail& other) const {
+  if (name != other.name) {
+    return name < other.name;
+  }
+
+  if (defaultKind != other.defaultKind) {
+    return defaultKind < other.defaultKind;
+  }
+
+  bool hasDecl = (decl != nullptr);
+  bool otherHasDecl = (other.decl != nullptr);
+  if (hasDecl != otherHasDecl) {
+    return hasDecl < otherHasDecl;
+  }
+  if (decl && other.decl && decl != other.decl) {
+    return decl->id() < other.decl->id();
+  }
+
+  if (isVarArgs != other.isVarArgs) {
+    return isVarArgs < other.isVarArgs;
+  }
+
+  return false;
+}
+
```

Comment:
Could you explain why all these `<`s are needed?

Reply:
It's for `TConverter::convertFunctionsToConvert`. That function converts all functions in a `CalledFnsSet` (`CalledFnsSet` is a set of functions = `std::unordered_map<const ResolvedFunction*, int>`). Because production AST does things like encode the order of declaration in the integer `id`; and because follow-on passes can depend on the ordering of AST declarations; we need this function to convert functions in the same order on every compilation, even though the `unordered_map` iteration order would be nondeterministic.